src/ray/gcs/redis_context.cc

```diff
@@ -598,55 +670,53 @@ Status RedisContext::Connect(const std::string &address,
+    enable_ssl_ = enable_ssl;
+    // Don't try to reconnect for the first time.
+    should_reconnect = false;
+  }
+
   // Fetch the ip address from the address. It might return multiple
   // addresses and only the first one will be used.
+  RAY_CHECK(!address.empty());
   auto ip_addresses = ResolveDNS(io_service_, address, port);
   RAY_CHECK(!ip_addresses.empty())
       << "Failed to resolve DNS for " << address << ":" << port;
 
   RAY_LOG(INFO) << "Resolve Redis address to " << absl::StrJoin(ip_addresses, ", ");
+  const auto &ip_address = ip_addresses[0];
 
-  {
-    auto resp = ConnectWithRetries<redisContext>(ip_addresses[0], port, redisConnect);
-    RAY_CHECK_OK(resp.first /* status */);
-    context_ = std::move(resp.second /* redisContext */);
-  }
-
-  if (enable_ssl) {
-    RAY_CHECK(ssl_context_ != nullptr);
-    RAY_CHECK(redisInitiateSSLWithContext(context_.get(), ssl_context_) == REDIS_OK)
-        << "Failed to setup encrypted redis: " << context_->errstr;
+  if (!ConnectToIPAddress(ip_address, port, username, password, enable_ssl)) {
+    if (!should_reconnect) {
+      RAY_LOG(FATAL) << "Failed to connect to Redis.";
+    }
+    return Reconnect();
   }
-  RAY_CHECK_OK(AuthenticateRedis(context_.get(), username, password));
 
-  // Connect to async context
-  std::unique_ptr<redisAsyncContext, RedisContextDeleter> async_context;
-  {
-    auto resp =
-        ConnectWithRetries<redisAsyncContext>(ip_addresses[0], port, redisAsyncConnect);
-    RAY_CHECK_OK(resp.first);
-    async_context = std::move(resp.second);
-  }
-  if (enable_ssl) {
-    RAY_CHECK(ssl_context_ != nullptr);
-    RAY_CHECK(redisInitiateSSLWithContext(&async_context->c, ssl_context_) == REDIS_OK)
-        << "Failed to setup encrypted redis: " << async_context->errstr;
-  }
-  RAY_CHECK_OK(AuthenticateRedis(async_context.get(), username, password));
-  redis_async_context_.reset(
-      new RedisAsyncContext(io_service_, std::move(async_context)));
-  SetDisconnectCallback(redis_async_context_.get());
+  // Check all contexts are connected before proceeding.
+  RAY_CHECK(context_ != nullptr || redis_async_context_ != nullptr)
+      << "Failed to connect to Redis.";
 
   // handle validation and primary connection for different types of redis
-  if (isRedisSentinel(*this)) {
-    return ConnectRedisSentinel(*this, username, password, enable_ssl);
-  } else {
-    return ConnectRedisCluster(*this,
-                               username,
-                               password,
-                               enable_ssl,
-                               ip_addresses[0] + ":" + std::to_string(port));
+  if (IsRedisSentinel()) {
+    return ConnectRedisSentinel(username, password, enable_ssl);
   }
+  return ConnectRedisCluster(username, password, enable_ssl);
+}
+
+Status RedisContext::Reconnect() {
+  RAY_LOG(INFO) << "Try to reconnect to Redis server.";
+  Disconnect();
+  return Connect(address_, port_, username_, password_, enable_ssl_);
```

Comment:
Are we going to have infinite reconnect if `address_` is down?

Reply:
Change it to fatal error we failed to connect to the saved address `RAY_REDIS_ADDRESS` in `Connect function.

Reply:
Let's do check failure first which matches the current behavior. Later on we can add retry different ip (since DNS may return multiple ips and currently we only try the first one).

Reply:
Does "later on" here mean implementing it in this PR or in a separate PR?

Reply:
Separate PR.